jgrapht-core/src/main/java/org/jgrapht/generate/RandomRegularGraphGenerator.java

```diff
@@ -0,0 +1,236 @@
+ * (b) the terms of the Eclipse Public License v1.0 as published by
+ * the Eclipse Foundation.
+ */
+
+package org.jgrapht.generate;
+
+import org.jgrapht.Graph;
+import org.jgrapht.VertexFactory;
+
+import java.util.*;
+
+/**
+ * Generate a random $d$-regular undirected graph with $n$ vertices.
+ * A regular graph is a graph where each vertex has the same degree, i.e. the same number of neighbors.
+ *
+ * <p>
+ * The algorithm for the simple case, proposed in [SW99] and extending the one for the non-simple case,
+ * runs in expected $\mathcal{O}(nd^2)$ time.
+ * It has been proved in [KV03] to sample from the space of random d-regular graphs
+ * in a way which is asymptotically uniform at random when $d = \mathcal{O}(n^{1/3 - \epsilon})$.
+ *
+ * <p>
+ * [KV03] Kim, Jeong Han, and Van H. Vu.
+ * "Generating random regular graphs."
+ * Proceedings of the thirty-fifth annual ACM symposium on Theory of computing. ACM, 2003.
+ *
+ * [SW99] Steger, Angelika, and Nicholas C. Wormald.
+ * "Generating random regular graphs quickly."
+ * Combinatorics, Probability and Computing 8.4 (1999): 377-396.
+ *
+ * @author Emilio Cruciani
+ * @since March 2018
+ *
+ * @param <V> graph node type
+ * @param <E> graph edge type
+ */
+public class RandomRegularGraphGenerator<V, E> implements GraphGenerator<V, E, V> {
+
+    private final int n;
+    private final int d;
+
+    /**
+     * Construct a new RandomRegularGraphGenerator.
+     *
+     * @param n number of nodes
+     * @param d degree of nodes
+     * @throws IllegalArgumentException if number of nodes is negative
+     * @throws IllegalArgumentException if degree is negative
+     * @throws IllegalArgumentException if degree is greater than number of nodes
+     * @throws IllegalArgumentException if the value "n * k" is odd
+     */
+    public RandomRegularGraphGenerator(int n, int d) {
+        if (n < 0) {
+            throw new IllegalArgumentException("number of nodes must be non-negative");
+        }
+        else if (d < 0) {
+            throw new IllegalArgumentException("degree of nodes must be non-negative");
+        }
+        else if (d > n) {
+            throw new IllegalArgumentException("degree of nodes must be smaller than or equal to number of nodes");
+        }
+        else if ((n * d) % 2 != 0) {
+            throw new IllegalArgumentException("value 'n * k' must be even");
+        }
+        else {
+            this.n = n;
+            this.d = d;
+        }
+    }
+
+    /**
+     * Generate a random regular graph.
+     *
+     * @param target target graph
+     * @param vertexFactory vertex factory
+     * @param resultMap result map
+     * @throws IllegalArgumentException if target is directed
+     * @throws IllegalArgumentException if "d == n" and graph is simple
+     */
+    @Override
+    public void generateGraph(Graph<V, E> target, VertexFactory<V> vertexFactory, Map<String, V> resultMap) {
+
+        // directed case
+        if (target.getType().isDirected()) {
```

Comment:
Better to write `!target.getType().isUndirected()`. We have a third case `isMixed()` and we might also add more in the future.